### PR TITLE
add Half support for layer_norm on CPU

### PR DESCRIPTION
**Pull request by jiayisunx**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99590

### Testing
Single socket (icx, 32cores):
| shape | fp32 forward (ms) | fp16 forward (ms) | mixed fp32 fp16 forward (ms) | fp32 backward (ms) | fp16 backward (ms) | mixed fp32 fp16 backward (ms) |
| -- | -- | -- | -- | -- | -- | -- |
| (1, 8, 16) | 0.012 | 0.011 | 0.011 | 0.051 | 0.051 | 0.050 |
| (8 ,8, 16) | 0.013 | 0.013 | 0.013 | 0.054 | 0.053 | 0.051 |
| (32, 8, 16) | 0.015 | 0.014 | 0.014 | 0.059 | 0.054 | 0.052 |
| (64, 128, 56, 56) | 1.875 | 0.790 | 1.016 | 12.845 | 7.151 | 6.985 |
| (64, 128, 256, 256) | 50.226 | 25.462 | 35.736 | 328.957 | 179.615 | 175.618 |

Single core (icx):

| shape | fp32 forward (ms) | fp16 forward (ms) | mixed fp32 fp16 forward (ms) | fp32 backward (ms) | fp16 backward (ms) | mixed fp32 fp16 backward (ms) |
| -- | -- | -- | -- | -- | -- | -- |
| (1, 8, 16) | 0.012 | 0.011 | 0.011 | 0.040 | 0.041 | 0.041 |
| (8 ,8, 16) | 0.012 | 0.012 | 0.012 | 0.042 | 0.042 | 0.042 |
| (32, 8, 16) | 0.027 | 0.014 | 0.014 | 0.048 | 0.048 | 0.046 |
| (64, 128, 56, 56) | 58.054 | 11.034 | 17.928 | 108.603 | 48.816 | 50.244 |
| (64, 128, 256, 256) | 1327.758 | 352.394 | 496.994 | 2846.182 | 1224.247 | 1218.422 |

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by mingfeima** [comment]
benchmark result?

File: aten/src/ATen/native/cpu/layer_norm_kernel.cpp [hunk 1]
```diff
@@ -23,14 +23,14 @@ namespace at::native {
...
-template <typename T, typename T_ACC>
+template <typename T>
 void LayerNormKernelImplInternal(
     const Tensor& X,
     const Tensor& gamma,
     const Tensor& beta,
     int64_t M,
     int64_t N,
-    T_ACC eps,
+    T eps,
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/layer_norm_kernel.cpp:33]: Is it expected to `eps` is in low precision?

File: test/test_mps.py [hunk 2]
```diff
@@ -125,6 +125,9 @@ def mps_ops_grad_modifier(ops):
 
         # round not working properly for float16
         'round': [torch.float16],
+
+        # Mixed datatype is recommended for layer_norm float16 on CPU
+        'nn.functional.layer_norm': [torch.float16],
```
Comment (jgong5) [inline test/test_mps.py:130]: Can you elaborate why we don't support it?

File: aten/src/ATen/native/cpu/layer_norm_kernel.cpp [hunk 3]
```diff
@@ -109,38 +111,40 @@ void layer_norm_kernel_mixed_type(
...
-        for (; d < N - (N % bVec::size()); d += bVec::size()) {
-          bVec x_bvec = bVec::loadu(X_ptr + d);
-          fVec x_fvec0, x_fvec1;
-          std::tie(x_fvec0, x_fvec1) = convert_bfloat16_float(x_bvec);
-          fVec gamma_fvec0, gamma_fvec1;
+      int64_t d = 0;
+      for (; d < N - (N % bVec::size()); d += bVec::size()) {
+        bVec x_bvec = bVec::loadu(X_ptr + d);
+        fVec x_fvec0, x_fvec1;
+        std::tie(x_fvec0, x_fvec1) = convert_to_float<T>(x_bvec);
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/layer_norm_kernel.cpp:126]: c++17: `auto [x_fvec0, x_fvec1]`

File: aten/src/ATen/native/cpu/layer_norm_kernel.cpp [hunk 4]
```diff
@@ -109,38 +111,40 @@ void layer_norm_kernel_mixed_type(
...
+        fVec gamma_fvec0, gamma_fvec1;
+        if (gamma_null) {
+          gamma_fvec0 = gamma_fvec1 = fVec(1);
+        } else {
           std::tie(gamma_fvec0, gamma_fvec1) = load2f(gamma_data + d);
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/layer_norm_kernel.cpp:131]: Would `auto [gamma_fvec0, gamma_fvec1] = gamma_null ? {fVec(1), fVec(1)} : load2f(gamma_data + d);` work?

File: aten/src/ATen/native/cpu/layer_norm_kernel.cpp [hunk 5]
```diff
@@ -109,38 +111,40 @@ void layer_norm_kernel_mixed_type(
...
-          y_bvec.store(Y_ptr + d);
         }
-        for (; d < N; d++) {
-          Y_ptr[d] = (X_ptr[d] * scale + bias) * gamma_data[d] + beta_data[d];
+        fVec beta_fvec0, beta_fvec1;
+        if (beta_null) {
+          beta_fvec0 = beta_fvec1 = fVec(0);
+        } else {
+          std::tie(beta_fvec0, beta_fvec1) = load2f(beta_data + d);
         }
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/layer_norm_kernel.cpp:138]: ditto

**Comment by jiayisunx** [comment]
> benchmark result?

I have added benchmark results, thanks!

**Review by mingfeima — APPROVED** [review]
so we have covered all the `reduce` utils from functional, right?

**Review by jgong5 — APPROVED** [review]

**Comment by jiayisunx** [comment]
> so we have covered all the `reduce` utils from functional, right?

Yes.

**Comment by jiayisunx** [comment]
@ngimel, can you please review this PR? thanks!

**Comment by jiayisunx** [comment]
@ngimel

**Comment by jiayisunx** [comment]
Hi @ngimel @mruberry @kulinseth , can you please review this PR? thanks!

**Comment by jiayisunx** [comment]
Hi @ngimel , can you please help review this PR? thanks!

**Comment by jiayisunx** [comment]
Hi @ngimel @mruberry @kulinseth , can you please help review this PR? thanks!

**Comment by CaoE** [comment]
@cpuhrsch Could you please review this PR ? Thank you.

**Comment by cpuhrsch** [comment]
@CaoE - Is this still current? Sorry, I was on vacation.

**Comment by CaoE** [comment]
@cpuhrsch Yes ~

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**